File: projects/typescript-examples/algokit-utils-ts/assets/asset-clawback.ts
import { setupLocalnetEnvironment } from '@/algokit-utils-ts/setup-localnet-environment'

async function assetClawback() {
  const { algorand, randomAccountA, randomAccountB } = await setupLocalnetEnvironment()

  // example: ASSET_CLAWBACK_TRANSACTION

  /**
   * An asset clawback transaction is an asset transfer transaction with the
   * `clawbackTarget` set to the account that is being clawed back from.
   *
   * Parameters for an asset transfer transaction:
   * - sender: The address of the account that will send the transaction
   * - assetId: ID of the asset
   * - amount: Amount of the asset to transfer (smallest divisible unit)
   * - receiver: The account to send the asset to
   * - clawbackTarget: The account to take the asset from, defaults to undefined
   */
  const txnResult = await algorand.send.assetTransfer({
    sender: randomAccountA.addr, // Must be the clawback address for the asset
    assetId: 1234n,
    amount: 1n,
    receiver: randomAccountA.addr,
    clawbackTarget: randomAccountB.addr, // account that is being clawed back from
  })

  console.log('Asset clawback transaction ID:', txnResult.transaction.txID)

  // example: ASSET_CLAWBACK_TRANSACTION
}

assetClawback().catch(console.error)

File: projects/python-examples/algokit_utils_py_examples/assets(ASA)/asset_clawback.py
from algokit_utils import AssetTransferParams

from algokit_utils_py_examples.helpers import (
    LocalnetEnvironment,
    setup_localnet_environment,
)


def asset_clawback() -> None:

    env: LocalnetEnvironment = setup_localnet_environment()
    algorand_client = env.algorand_client
    manager = env.account_a  # Using account_a as the manager
    account_to_be_clawbacked = (
        env.account_b
    )  # Using account_b as the account to be clawed back

    # example: ASSET_CLAWBACK_TRANSACTION

    """
    An asset clawback transaction is an asset transfer transaction with the
    `clawback_target` set to the account that is being clawed back from.

    Parameters for an asset transfer transaction.
    - sender: The address of the account that will send the transaction
    - asset_id: ID of the asset
    - amount: Amount of the asset to transfer (smallest divisible unit)
    - receiver: The account to send the asset to
    - clawback_target: The account to take the asset from, defaults to None
    """

    txn_result = algorand_client.send.asset_transfer(
        AssetTransferParams(
            sender=manager.address,
            asset_id=1234,
            amount=1,
            receiver=manager.address,
            clawback_target=account_to_be_clawbacked.address,  # account that is being clawed back from
        )
    )

    # example: ASSET_CLAWBACK_TRANSACTION
